core/src/main/scala/filodb.core/reprojector/FiloMemTable.scala

```diff
@@ -0,0 +1,216 @@
+
+import com.typesafe.config.Config
+import java.nio.ByteBuffer
+import java.util.concurrent.{Executors, ScheduledFuture, TimeUnit}
+import java.util.TreeMap
+import net.ceedubs.ficus.Ficus._
+import org.velvia.filo.{VectorInfo, RowToVectorBuilder, RowReader, FiloRowReader, FastFiloRowReader}
+import scala.collection.mutable.{ArrayBuffer, HashMap}
+import scala.concurrent.duration.FiniteDuration
+import scala.math.Ordered
+import scalaxy.loops._
+
+import filodb.core.{KeyRange, SortKeyHelper}
+import filodb.core.Types._
+import filodb.core.columnstore.{RowWriterSegment, Segment}
+import filodb.core.metadata.{Column, Dataset, RichProjection}
+
+/**
+ * A MemTable using Filo vectors to store rows in memory, plus an index to seek into the chunks.
+ * The index is just an on-heap TreeMap per partition and keeps rows in sorted order.
+ * The idea is to minimize serialization costs by leveraging Filo vectors, as compared to MapDB,
+ * which has to do several expensive key serialization steps when inserting into a Map.
+ * Reads are still efficient because Filo vectors are designed for fast random access and minimal
+ * deserialization.
+ * New rows are kept in tempRows while enough rows or time elapses to be flushed into columnar
+ * chunks.  If not enough rows before flush.interval elapses, then the chunk is flushed anyways, but will be
+ * appended to next time, such that each chunk always fills up.
+ *
+ * ==Config==
+ * {{{
+ *   memtable {
+ *     filo.chunksize = 1000   # The number of rows per Filo chunk
+ *     flush.interval = 1 s    # If less than chunksize rows ingested before this interval, then
+ *                             # rows will be flushed out anyways
+ *   }
+ * }}}
+ */
+class FiloMemTable[K](val projection: RichProjection[K], config: Config) extends MemTable[K] {
+  import RowReader._
+  import collection.JavaConverters._
+
+  val chunkSize = config.as[Option[Int]]("memtable.filo.chunksize").getOrElse(1000)
+  val flushInterval = config.as[FiniteDuration]("memtable.flush.interval")
+
+  // From sort key K to a Long: upper 32-bits = chunk index, lower 32 bits = row index
+  type KeyMap = TreeMap[K, Long]
+
+  private val chunks = new ArrayBuffer[Array[ByteBuffer]]
+  private val readers = new ArrayBuffer[FiloRowReader]
+  private val partKeyMap = new HashMap[PartitionKey, KeyMap]
+
+  private val filoSchema = projection.columns.map {
+    case Column(name, _, _, colType, serializer, false, false) =>
+      require(serializer == Column.Serializer.FiloSerializer)
+      VectorInfo(name, colType.clazz)
+  }
+  private val clazzes = filoSchema.map(_.dataType).toArray
+  private val colIdToNumber =
+    projection.columns.zipWithIndex.map { case (col, idx) => (col.name, idx) }.toMap
+
+  // Holds temporary rows before being flushed to chunk columnar storage
+  private val tempRows = new ArrayBuffer[RowReader]
+  private val callbacks = new ArrayBuffer[(Int, Int, () => Unit)]   // Start and end row index for each callback
```

Comment:
Hmmm ... the synchronized is used on an ongoing basis, because scheduled flushes happen on a different thread, and potentially at the same time as an insert which modifies the tempRows and callbacks.